client.go

```diff
@@ -306,11 +306,65 @@ func (c *client) unassignRoles(ctx context.Context, roleIDs []string) error {
 	return merr.ErrorOrNil()
 }
 
+// addGroupMemberships adds the service principal to the Azure groups.
+func (c *client) addGroupMemberships(ctx context.Context, sp *graphrbac.ServicePrincipal, groups []*AzureGroup) ([]string, error) {
+	groupsIDs := make([]string, 0, len(groups))
+	for _, group := range groups {
+		_, err := retry(ctx, func() (interface{}, bool, error) {
+			_, err := c.provider.AddGroupMember(ctx, group.ObjectID,
+				graphrbac.GroupAddMemberParameters{
+					URL: to.StringPtr(
+						fmt.Sprintf("https://graph.windows.net/%s/directoryObjects/%s",
+							c.settings.TenantID,
+							*sp.ObjectID,
+						),
+					),
+				})
+
+			// Propagation delays within Azure can cause this error occasionally, so don't quit on it.
+			if err != nil && strings.Contains(err.Error(), "Request_ResourceNotFound") {
+				return nil, false, nil
+			}
+
+			groupsIDs = append(groupsIDs, group.ObjectID)
+			return nil, true, err
+		})
+
+		if err != nil {
+			return nil, errwrap.Wrapf("error while adding group membership: {{err}}", err)
+		}
+	}
+
+	return groupsIDs, nil
+}
+
+// removeGroupMemberships removes the passed service principal from the passed
+// groups. This is a clean-up operation that isn't essential to revocation. As
+// such, an attempt is made to remove all memberships, and not return
+// immediately if there is an error.
+func (c *client) removeGroupMemberships(ctx context.Context, servicePrincipalID string, groupIDs []string) error {
```

Comment:
Why pass in a string instead of `*graphrbac.ServicePrincipal` here? (re: consistency)

Reply:
Oh, I see this is just from internal data strings, so that's probably fine. Though getting the word "Object" into the parameter name would be good, since ApplicationID vs. ObjectID is always error prone.